File: Practice/DFS_BFS/q21/q21.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <cstring>

using namespace std;

int n, l, r;
int board[50][50];
int chk[50][50];
int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};
vector<pair<int, int>> u;

bool check_diff(int diff){
    return diff >= l && diff <= r;
}

int bfs(int x, int y){
    int population = 0, count = 0;
    queue<pair<int,int>> q;
    q.push(make_pair(x, y));
    chk[x][y] = 1;
    while(!q.empty()){
        pair<int,int> root = q.front();
        u.push_back(make_pair(root.first, root.second));
        population += board[root.first][root.second];
        count++;
        q.pop();
        for(int i = 0; i < 4; i++){
            int nx = root.first + dx[i];
            int ny = root.second + dy[i];
            if(nx >= 0 && nx < n && ny >= 0 && ny < n && chk[nx][ny] == 0){
                int diff = abs(board[root.first][root.second]-board[nx][ny]);
                if(check_diff(diff)){
                    q.push(make_pair(nx, ny));
                    u.push_back(make_pair(nx, ny));
                    chk[nx][ny] = 1;
                }
            }
        }
    }
    return population/count;
}

bool update_population(int population){
    int num = u.size();
    for(int i = 0; i < num; i++){
        int x = u[i].first;
        int y = u[i].second;
        board[x][y] = population;
    }
    u.clear();
    return num != 1;
}

int set_union(){
    int count = 0;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            if(chk[i][j] == 0){
                if(update_population(bfs(i, j))){
                    count++;
                }
            }
        }
    }
    memset(chk, 0, sizeof(chk));
    return count;
}

int main(void){
    cin >> n >> l >> r;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            cin >> board[i][j];
        }
    }
    int ans = 0;
    while(set_union() != 0) ans++;
    cout << ans << "\n";
}

File: Practice/DFS_BFS/q21/q21.py
from collections import deque

n, l, r = map(int, input().split())
board = []
direction = [(-1, 0), (0, 1), (0, -1), (1, 0)]
chk = [[False for _ in range(n)] for _ in range(n)] 
for _ in range(n):
    board.append(list(map(int, input().split())))

def check_union(p1, p2):
    diff = abs(p1-p2)
    return diff >= l and diff <= r

def bfs(x, y):
    total = 0
    union = []
    queue = deque()
    queue.append((x, y))
    chk[x][y] = True
    while len(queue) != 0:
        c = queue.popleft()
        union.append(c)
        total += board[c[0]][c[1]]
        for d in direction:
            nx = c[0] + d[0]
            ny = c[1] + d[1]
            if nx >= 0 and nx < n and ny >= 0 and ny < n and not chk[nx][ny] and check_union(board[c[0]][c[1]], board[nx][ny]):
                queue.append((nx, ny))
                chk[nx][ny] = True
    total //= len(union)
    for c in union:
        board[c[0]][c[1]] = total
    return 1 if len(union) != 1 else 0

def set_union():
    global chk
    union_count = 0
    for i in range(n):
        for j in range(n):
            if chk[i][j] == False:
                union_count += bfs(i, j)
    chk = [[False for _ in range(n)] for _ in range(n)] 
    return union_count

ans = 0
while set_union() != 0:
    ans += 1

print(ans)